File: src/arduinobot_cpp_examples/src/simple_lifecycle_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>
#include <std_msgs/msg/string.hpp>
#include <string>
#include <thread>

using std::placeholders::_1;
using namespace std::chrono_literals;

class SimpleLifecycleNode : public rclcpp_lifecycle::LifecycleNode
{
public:
    explicit SimpleLifecycleNode(const std::string &node_name, bool intra_process_comms = false) : LifecycleNode(node_name, rclcpp::NodeOptions().use_intra_process_comms(intra_process_comms))
    {
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_configure(const rclcpp_lifecycle::State &)
    {
        subscriber_ = create_subscription<std_msgs::msg::String>("example_topix", 10, std::bind(&SimpleLifecycleNode::OnMessageReceived, this, _1));
        RCLCPP_INFO(get_logger(), "simple_lifecycle_node on_configure() executed");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_shutdown(const rclcpp_lifecycle::State &)
    {
        subscriber_.reset();
        RCLCPP_INFO(get_logger(), "simple_lifecycle_node on_shutdown() executed");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_cleanup(const rclcpp_lifecycle::State &)
    {
        subscriber_.reset();
        RCLCPP_INFO(get_logger(), "simple_lifecycle_node on_cleanup() executed");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_activate(const rclcpp_lifecycle::State & state)
    {
        LifecycleNode::on_activate(state);
        std::this_thread::sleep_for(2s);
        RCLCPP_INFO(get_logger(), "simple_lifecycle_node on_activate() executed");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_deactivate(const rclcpp_lifecycle::State & state)
    {
        LifecycleNode::on_deactivate(state);
        RCLCPP_INFO(get_logger(), "simple_lifecycle_node on_deactivate() executed");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    void OnMessageReceived(const std_msgs::msg::String::SharedPtr msg) const
    {
        auto state = get_current_state();
        if(state.label() == "active")
            RCLCPP_INFO_STREAM(get_logger(), "Received message: " << msg->data.c_str());
    }

private:
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::executors::SingleThreadedExecutor ste;
    auto simpleLifecycleNode = std::make_shared<SimpleLifecycleNode>("simple_lifecycle_node");
    ste.add_node(simpleLifecycleNode->get_node_base_interface());
    ste.spin();
    rclcpp::shutdown();
    return 0;
}

File: src/arduinobot_py_examples/arduinobot_py_examples/simple_lifecycle_node.py
import time
import rclpy
import rclpy.executors
from rclpy.lifecycle import Node, State, TransitionCallbackReturn
from std_msgs.msg import String


class SimpleLifeCycleNode(Node):
    def __init__(self, node_name, **kwargs):
        super().__init__(node_name, **kwargs)

    def on_configure(self, state: State) -> TransitionCallbackReturn:
        self.subcriber_ = self.create_subscription(
            String, "example_topix", self.msgCallback, 10
        )
        self.get_logger().info('simple_lifecycle_node on_configure() executed')
        return TransitionCallbackReturn.SUCCESS

    def on_cleanup(self, state: State) -> TransitionCallbackReturn:
        self.destroy_subscription(self.subcriber_)
        self.get_logger().info('simple_lifecycle_node on_cleanup() executed')
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('simple_lifecycle_node on_activate() executed')
        time.sleep(2)
        return super().on_activate(state)

    def on_deactivate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('simple_lifecycle_node on_deactivate() executed')
        return super().on_deactivate(state)

    def msgCallback(self, msg):
        if self._state_machine.current_state[1] == "active":
            self.get_logger().info(f'Received message: {msg.data}')


def main(args=None):
    rclpy.init()
    executor = rclpy.executors.SingleThreadedExecutor()
    simpleLifeCycleNode = SimpleLifeCycleNode("simple_lifecycle_node")
    executor.add_node(simpleLifeCycleNode)
    try:
        executor.spin()
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        simpleLifeCycleNode.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()